feat: add new parser with fixes and improvements

obycode opened the pull request:
### Description
This PR adds an updated parser which fixes several issues and improves performance. 

### Applicable issues
- fixes #3124, #3123, #3128

### Additional info (benefits, drawbacks, caveats)
Use of the new parser is gated on epoch 2.1 and the old parser is used prior to then.

### Checklist
- [x] Test coverage for new or modified code paths
- [x] Changelog is updated
- [x] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

kantai reviewed (approved):
This looks great to me (and very much an improvement over the v1 parser)! Thanks @obycode!

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/error.rs:
```diff
@@ -0,0 +1,79 @@
+            SingleSemiColon => "unexpected single ';' (comments begin with \";;\"".to_string(),
+            UnknownSymbol(c) => format!("unknown symbol, '{}'", c),
+            NoteToMatchThis(c) => format!("to match this '{}'", c),
+            UnsupportedLineEnding => {
```
Do these errors show up in the blockchain's event stream?  These two errors might be confusing in that context.

obycode replied:
No, these are `Note` level diagnostics, so they are ignored when in `fail_fast` mode.

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,1957 @@
+
+fn is_separator(ch: char) -> bool {
+    match ch {
+        '(' | ')' | '{' | '}' | ',' | ':' | '\0' | '.' => true,
```
What does `\0` separate in v2?  I don't think it's even legal in v1?

obycode replied:
It should not be allowed in the contract. It's just used as a placeholder for the end of the contract string (see line 82 below):

```rust
        match self.input.next() {
             Some(ch) => {
                 if self.next == '\n' {
                     self.line = self.line + 1;
                     self.column = 0;
                 }
                 self.next = ch
             }
             None => self.next = '\0',
         }
```

jcnelson replied:
That's the thing -- I don't think strings in rust are guaranteed to be null-terminated.  There isn't a null byte that terminates the contract payload, for example.

What are we doing to be sure that there is absolutely no way that `\0`, or any other ASCII control code besides `\t` and `\n`, can show up in the code body?  Is this handled by a regex somewhere before the contract is fed into the lexer?

obycode replied:
Ah, thanks, I see the problem. This technique is not relying on a null-terminator in the rust string, but it is missing a check for `\0` showing up in the code. If it does show up, it would be treated like the end of the file and cause a problem. I'll fix that.

obycode replied:
Fixed in dbb2664fb

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,1957 @@
+        loop {
+            self.read_char();
+            match self.next {
+                'A'..='Z' | '0'..='9' => principal.push(self.next),
```
Principals use the Crockford 32 alphabet, which is `0123456789ABCDEFGHJKMNPQRSTVWXYZ` (note that `I`, `L`, and `O` are _not_ in this alphabet).

obycode replied:
Fixed in 2d825c5ba

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,1957 @@
+        }
+    }
+
+    fn read_unsigned(&mut self) -> u128 {
```
I don't think this code does what it's supposed to do if there is ever a literal unsigned number that is greater than `u128::MAX`.  If this happens, then the whole contract should fail analysis.  It looks like this code instead attempts to convert the textual representation into a `u128` without handling the overflow case.

jcnelson replied:
I think such an occurence should cause the lexer to log an error at this location?

jcnelson replied:
More of a design-level question: I see that the lexer will return a `String` for every other token, and rely on the parser to try to convert the `String` into a Clarity `Value` of the corresponding type.  The parser, not the lexer, identifies malformed tokens.  Should that also be happening here, for `uint` and `int`?

obycode replied:
Yes, that makes sense. The lexer should just be verifying the allowed characters are used and the parser should validate the complete token. I will make that change.

obycode replied:
Fixed in a43454bd2

obycode replied:
Rebased, now 639e4030e348158f63c0b8fbacbb52083b95fb74

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,1957 @@
+        num
+    }
+
+    fn read_integer(&mut self) -> i128 {
```
I have the same concern here as with `read_unsigned`.  What happens if there is a literal that requires more than 128 bits to represent?

obycode replied:
Fixed in a43454bd2

obycode replied:
Rebased, now 639e4030e348158f63c0b8fbacbb52083b95fb74

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,1957 @@
+        }
+    }
+
+    fn read_utf8_encoding(&mut self) -> String {
```
I think this is meant to read `X` in `\u{X}`, right?  If so, can you document this?

obycode replied:
Added a comment in e5d99b691a270764a7b55e09ed40825cb765d3e4

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,1957 @@
+                    'u' => {
+                        let encode_str = self.read_utf8_encoding();
+                        let unicode_char = {
+                            let u = u32::from_str_radix(&encode_str, 16).unwrap();
```
What if `encode_str` is more than 8 characters long, and thus can't fit into `u32`?

obycode replied:
Fixed in cda9ec15c05757dc68f4ec1a2614cc53e8177817

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,1957 @@
+        }
+    }
+
+    fn read_principal(&mut self) -> String {
```
Ah, I think this code just reads `X` in `'X`, right?  It doesn't really care whether or not `X` represents a valid principal?

obycode replied:
Yes, this is just reading `X` in `'X`.

Right, the parser checks that using `parse_standard_principal`, but I had not realized that principals use the Crockford 32 alphabet. Maybe the lexer should catch that, since it would be able to give a more specific error in the case that an illegal character is used. I will modify the checks there.

obycode replied:
Fixed in 2d825c5ba

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,1957 @@
+                self.read_char();
+                if self.next == 'x' {
+                    Token::Bytes(self.read_hex())
+                } else if self.next.is_ascii_digit() {
```
Wait, what about the string "0)"?  Such as in `(print (< x 0))`?

obycode replied:
That would reach the next case,

```rust
                } else if is_separator(self.next) {
                    Token::Int("0".to_string())
                }
```

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,1957 @@
+        assert_eq!(lexer.diagnostics.len(), 0);
+
+        lexer = Lexer::new("\"\\r\\t\\0\\\\ ok\"", false);
+        assert_eq!(
```
I don't think this is allowed today in Clarity -- you can't have bytes`\r` or `\0` in the code body at all.

obycode replied:
Inside a string, they are allowed in the current parser, though I am noticing now that the `\0` will later trigger an error (the `\r` is allowed) from:

https://github.com/stacks-network/stacks-blockchain/blob/38aa968d7e81d575453ba5253aa8e64a8ac7cec9/clarity/src/vm/types/mod.rs#L767-L771

So perhaps instead we should just disallow `\0` in the parser? And also remove `\r` if we don't want it to be supported.

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,1957 @@
+
+        loop {
+            match self.next {
+                'a'..='z' | 'A'..='Z' | '0'..='9' | '_' | '-' => ident.push(self.next),
```
Ah, never mind, it seems they do not.

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,2518 @@
+            return None;
+        }
+        let token = self.tokens[self.next_token].clone();
+        self.next_token = self.next_token + 1;
```
This is also legal rust :)

```rust
self.next_token += 1;
```

obycode replied:
Oops 😆. d00ff689f

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,2518 @@
+                nodes.push(node);
+                whitespace = self.ignore_whitespace();
+            } else {
+                let token = self.tokens[self.next_token - 1].clone();
```
We're sure that `self.next_token > 0` here, always?

obycode replied:
Yes, we know for sure that `self.next_token > 0` because the only way to get into `parse_list` is to have already processed a `(`, which would increment the count. I checked other locations where 1 is subtracted as well.

We also have a guarantee that there is always at least one token from the lexer, the `EOF` indicator, so even in the case of an empty string, one token will be processed successfully before reaching the end. This makes the `self.next_token - 1` in `parse_node_or_eof` safe.

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,2518 @@
+                    // Create a placeholder value so that parsing can continue,
+                    // then return.
+                    self.add_diagnostic(ParseErrors::TupleColonExpectedv2, token.span.clone())?;
+                    let mut placeholder =
```
Should there be a dedicated `PreSymbolicExpression` type variant for placeholders like this?

obycode replied:
Yes, there should be. When I first designed this, I was avoiding making changes there, but you're right, at this point it would be the right move. 

I also intended to create a type for comments, because there are many things that the tools could do if we maintained the comments in the AST instead of just treating them like whitespace. Are you okay with me adding that into this PR as well if I will already be making changes there?

jcnelson replied:
I'm personally fine with it since I don't think the change will be very big, but let's also ask @kantai.  He knows this part of the codebase better than I do.

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,2518 @@
+            return Ok(None);
+        }
+        let token = token.unwrap();
+        // let mut comment = None;
```
dead code :)

obycode replied:
Removed

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,2518 @@
+                            ParseErrors::IllegalASCIIString(val),
+                            token.span.clone(),
+                        )?;
+                        Value::string_ascii_from_bytes("placeholder".as_bytes().to_vec()).unwrap()
```
Should there be a `Value::Placeholder` type variant instead of a bespoke value for `Value` that indicates as such?

obycode replied:
Yes, added.

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,2518 @@
+                Some(e)
+            }
+            Token::TraitIdent(mut name) => {
+                if name.len() > MAX_STRING_LEN {
```
Ah, nm, I see this only applies to the trait identifier from `(use-trait ...)`, which must be a `ClarityName`

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,2518 @@
+                Some(e)
+            }
+            Token::Principal(addr) => {
+                let principal = match PrincipalData::parse_standard_principal(&addr) {
```
I see that this is where the check that the `addr` is a well-formed address comes in.  However, if the user creates a principal literal that includes an invalid Crockford 32 character, they'll get `ParseErrors::InvalidPrincipalLiteral` instead of `LexerError::InvalidCharPrincipal`.  Is this the desired behavior?

obycode replied:
Updated in 2d825c5 to now generate the `InvalidCharPrincipal`.

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,2518 @@
+        match &exprs[3].pre_expr {
+            PreSymbolicExpressionType::AtomValue(Value::Sequence(SequenceData::String(
+                CharType::ASCII(ASCIIData { data: s }),
+            ))) => assert_eq!(s, "hel\tlo".as_bytes()), //assert_eq!(val.expect_ascii(), "hel\tlo"),
```
Dead code :)

obycode replied:
Removed in a8a3ff390

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,2518 @@
+                end_line: 1,
+                end_column: 137
+            }
+        );
```
Can you add a test for a principal with invalid Crockford 32 characters?

obycode replied:
Added in 621f4db9d

jcnelson reviewed (changes requested):
I have a few comments that need to be addressed before I'm okay with merging this.  I think there may be some errors in the implementation.

obycode commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,1957 @@
+                    'n' => '\n',
+                    't' => '\t',
+                    'r' => '\r',
+                    '0' => '\0',
```
Yes, this is the same set from the old parser:

https://github.com/stacks-network/stacks-blockchain/blob/38aa968d7e81d575453ba5253aa8e64a8ac7cec9/clarity/src/vm/ast/parser/mod.rs#L436-L446

I can add others if we decide we need to support others.

obycode commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,1957 @@
+                        return error_string.to_string();
+                    }
+                    self.read_char();
+                    return code;
```
Sticking with the rule of only validating characters in the lexer and checking higher-level properties in the parser, I think this can stay as-is.

obycode replied:
This is now actually checked in the caller, `read_utf8_string`. Fixed in cda9ec15c05757dc68f4ec1a2614cc53e8177817.

obycode commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,2518 @@
+                e.span = token.span;
+                Some(e)
+            }
+            // TODO: For now, comments are being treated as whitespace. In the future,
```
Based on the discussion above, I will add this feature and include it in this PR.

obycode commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,2518 @@
+                end_line: 1,
+                end_column: 5
+            }
+        );
```
Yes, all of these variants are now legal. Tests added in 29486c63d.

obycode commented:
Thanks for the review Jude. I've addressed most of the comments and pushed updates to the branch. The remaining items are to add `PreSymbolicExpression` type variants for comments and placeholders.

I just noticed the conflict, so I will also rebase to resolve that.

obycode commented:
a1a7ffe has two big improvements based on your feedback:
* Adds the placeholder type
  - Used to continue parsing through errors
  - Holds the original string, so can be used to reproduce the source
* Refactors the lexer
  - All lexer tokens are representations of the original string, leaving the further processing for the parser
  - UTF8 strings are a slight rule bender, as the utf8 code points are validated in the lexer and the decoded utf8 string is passed to the parser

I also added the comment type, but it is not used yet. I think after that change to handle comments, it will be ready for re-review. I will also review the code coverage and add new tests as needed after this last change.

obycode commented:
All items should be addressed now, other than the one question about whether we want to change the allowance of `\0` and `\r` characters.

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,2207 @@
+
+fn is_string_terminator(ch: char) -> bool {
+    match ch {
+        '"' | '\n' | EOF => true,
```
Does this mean that the following contract will parse?

```clarity
"hello world
```

I think you can have "bare" contracts that don't have any lists, but just raw literals.  I just want to make sure that calling `EOF` a string terminator doesn't lead to silliness like this.

obycode replied:
No, this is just used to read through the "end" of a string when an error is encountered. If a `"` is not found before the end of the line or the end of the file, then the string is considered closed. This is only reached when not in "fail fast" mode, when attempting to continue parsing past an error.

jcnelson commented on clarity/src/vm/ast/parser_v2/lexer/mod.rs:
```diff
@@ -0,0 +1,2207 @@
+        lexer = Lexer::new("0123", false).unwrap();
+        assert_eq!(
+            lexer.read_token().unwrap().token,
+            Token::Int("123".to_string())
```
Wait, shouldn't this be `Token::Int("123".to_string())`?  Why does the lexer strip leading 0's?  Isn't that the parser's job?

obycode replied:
Yes, good catch. This should have changed with the switch making the lexer return strings for integers. Fixed in 4851c3880.

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,3227 @@
+            }
+        };
+
+        let contract_name = match ContractName::try_from(name.clone()) {
```
Should this code be checking that the contract name's length isn't too long here?  This isn't done in `ContractName::try_from()`.

obycode replied:
Fixed and tested in 9d2634ef9

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,3227 @@
+    (stmts, diagnostics, parser.success)
+}
+
+pub fn decode_hex(s: &str) -> Result<Vec<u8>, ParseIntError> {
```
This duplicates `stacks_common::util::hash::hex_bytes`.  Can you use that method instead?

obycode replied:
Great! Yes, removed in 9324482ec.

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,3227 @@
+
+        let (stmts, diagnostics, success) =
+            parse_collect_diagnostics("340282366920938463463374607431768211456 ");
+        assert_eq!(success, false);
```
This fails due to numeric overflow, right?

obycode replied:
Yes, this number exceeds the range of a 128-bit integer. I'll add a comment.

obycode replied:
5525fcead

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,3227 @@
+                end_line: 1,
+                end_column: 39
+            }
+        );
```
Does the parser correctly handle `"0000000000123"` an `"u00000000123"`?

obycode replied:
Do you mean `"0000000000123"` with the quotes, as a string, or `0000000000123` as an integer? Both are handled correctly, but I added tests for the int/uint in a52eef3c5.

jcnelson commented on clarity/src/vm/ast/parser_v2/mod.rs:
```diff
@@ -0,0 +1,3227 @@
+        );
+
+        let (stmts, diagnostics, success) = parse_collect_diagnostics("{1: u2, 3: u4}");
+        assert_eq!(success, true);
```
Wait, shouldn't this fail? Keys of a tuple can't be numeric, can they?

obycode replied:
The parser allows this, but the type-checker reports an error.

jcnelson commented on clarity/src/vm/ast/traits_resolver/mod.rs:
```diff
@@ -176,24 +177,33 @@ impl TraitsResolver {
     fn probe_for_generics(
         &mut self,
-        exprs: &[PreSymbolicExpression],
+        exprs: Vec<&PreSymbolicExpression>,
```
Why is his now a `Vec`?

jcnelson reviewed (approved):
This LGTM now.  Thanks @obycode!

obycode commented:
I just completed some more targeted timing of the parsers and it showed that this v2 parser is more than 8x faster than the v1 parser!

obycode commented:
@jcnelson I just noticed that the unit tests failed due to a seemingly random error from codecov, so I triggered a re-run of those tests. Please help me merge after that testing is complete.

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
